cmd/chartsvc/handler.go

```diff
@@ -61,34 +64,125 @@ type rel struct {
 	Links selfLink    `json:"links"`
 }
 
-// listCharts returns a list of charts
-func listCharts(w http.ResponseWriter, req *http.Request) {
+type meta struct {
+	TotalPages int `json:"totalPages"`
+}
+
+// getPageNumberAndSize extracts the page number and size of a request. Default (1, 0) if not set
+func getPageNumberAndSize(req *http.Request) (int, int) {
+	page := req.FormValue("page")
+	size := req.FormValue("size")
+	pageInt, err := strconv.ParseUint(page, 10, 64)
+	if err != nil {
+		pageInt = 1
+	}
+	// ParseUint will return 0 if size is a not positive integer
+	sizeInt, _ := strconv.ParseUint(size, 10, 64)
+	return int(pageInt), int(sizeInt)
+}
+
+// min returns the minimum of two integers.
+// We are not using math.Min since that compares float64
+// and it's unnecesary complex.
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func uniqChartList(charts []*models.Chart) []*models.Chart {
+	// We will keep track of unique digest:chart to avoid duplicates
+	chartDigests := map[string]bool{}
+	res := []*models.Chart{}
+	for _, c := range charts {
+		digest := c.ChartVersions[0].Digest
+		// Filter out the chart if we've seen the same digest before
+		if _, ok := chartDigests[digest]; !ok {
+			chartDigests[digest] = true
+			res = append(res, c)
+		}
+	}
+	return res
+}
+
+func getPaginatedChartList(repo string, pageNumber, pageSize int) (apiListResponse, interface{}, error) {
 	db, closer := dbSession.DB()
 	defer closer()
 	var charts []*models.Chart
-	if err := db.C(chartCollection).Find(nil).Sort("name").All(&charts); err != nil {
+
+	c := db.C(chartCollection)
+	pipedActions := []bson.M{}
+	var countQuery datastore.Query
+	if repo != "" {
+		pipedActions = append(pipedActions, bson.M{"$match": bson.M{"repo.name": repo}})
+		countQuery = c.Find(bson.M{"repo.name": repo})
+	} else {
+		countQuery = c.Find(nil)
+	}
+
+	// We should query unique charts
+	pipedActions = append(pipedActions,
+		// Add a new field to store the latest version
+		bson.M{"$addFields": bson.M{"firstChartVersion": bson.M{"$arrayElemAt": []interface{}{"$chartversions", 0}}}},
+		// Group by unique digest for the latest version (remove duplicates)
+		bson.M{"$group": bson.M{"_id": "$firstChartVersion.digest", "chart": bson.M{"$first": "$$ROOT"}}},
+		// Restore original object struct
+		bson.M{"$replaceRoot": bson.M{"newRoot": "$chart"}},
+		// Order by name
+		bson.M{"$sort": bson.M{"name": 1}},
+	)
+
+	totalPages := 1
+	if pageSize != 0 {
+		// If a pageSize is given, returns only the the specified number of charts and
+		// the number of pages
+		length, err := countQuery.Count()
```

Comment:
You are right, I missed that. The good news is that the repeated pipeline steps are cached so after executing "count" the retrieval of charts is faster.

To be fair, I have rechecked the times because I think this is getting a bit complex. The aggregation solution is slightly slower than querying everything, picking unique charts and do the pagination in the chartsvc (40 vs 50 ms). The good part of doing aggregation is that we need to maintain less code in our side, the downside is that we increase the vendor lock-in with MongoDB.

Reply:
> To be fair, I have rechecked the times because I think this is getting a bit complex. The aggregation solution is slightly slower than querying everything, picking unique charts and do the pagination in the chartsvc (40 vs 50 ms). The good part of doing aggregation is that we need to maintain less code in our side (with the caveat that we need the code to filter unique charts anyway), the downside is that we increase the vendor lock-in with MongoDB.

This is a good analysis. One thing to consider is how this might scale if we had 1,000 charts, or 10, 000 - my assumption is MongoDB will be faster as it avoids having to send all those charts down the pipe to the chartsvc. There might also be some ways to improve the performance of the aggregation by configuring indexes in MongoDB.

For the same reasons as above, if we do end up switching to a different database in the future, we will likely want to reimplement the pipeline as a query for this new database (e.g. some sort of distinct SQL query if we go with a relational database). That said, this probably won't happen anytime soon.

Another solution we haven't considered yet is doing this filtering in `chart-repo`, though race conditions would make that more difficult (possibly solvable using transactions, though). Not saying we should consider this now, but worth keeping in mind.